Feature/app auth

darrenleeweber opened the pull request:
This PR is for review of the current state of the auth_controller, which is or should be close to the requirements of the IIIF image authentication, see http://image-auth.iiif.io/api/image/2.1/authentication.html.  (The main exception to that spec is missing support for JSONP requests.)

Further work to adapt this for Triannon will be done in another branch:
  - https://github.com/sul-dlss/triannon/tree/feature/auth-ish

ndushay commented:
BTW, it's possible to compare branches in github without creating a PR (but I don't think you can create the artifact - only do it live)

darrenleeweber commented:
I'm proposing a merge of this feature into the develop branch, to pursue the next phase of this auth development, by issuing a PR for the auth-ish branch.  Unless I hear otherwise, I'll proceed with this merge on Thu or Fri (June 18-19).

darrenleeweber commented:
@ndushay - this PR is ready for review, as I think the coveralls failure is trivial and, if not, it looks like something for you to review (I've taken care of things that I've modified, I think).

darrenleeweber commented:
@azaroth42 - this PR represents an integration of the multi-roots with a "IIIF-spec" for auth.  Once we merge this PR, additional revisions to auth will be in the 'auth-ish' branch (and eventual PR).

ndushay commented on app/controllers/triannon/auth_controller.rb:
```diff
@@ -0,0 +1,353 @@
+    include RdfResponseFormats
+
+    AUTH_EXPIRY  =   60 # seconds
+    TOKEN_EXPIRY = 3600 # seconds
```
would it be more appropriate to allow expiry times to be chosen in the triannon.yml  file?

darrenleeweber replied:
@azaroth42 - please consider whether these auth requirements for timeouts are fixed values that can be hard-coded or variable values that could and should be config parameters?  As I've interpreted it, they are fixed values that are not negotiable and @ndushay raises a good question about whether that should be so.

azaroth42 replied:
:+1: to having them configurable params. 60/3600 are probably fine for most use cases, but I'm sure there'll be situations when they're not. Thanks ND, DW!

ndushay commented on app/controllers/triannon/auth_controller.rb:
```diff
@@ -0,0 +1,353 @@
+    # http://image-auth.iiif.io/api/image/2.1/authentication.html#error-conditions
+    # return json body [String] containing: { "authorizationCode": code }
+    def client_identity
+      unless request.post?
```
I believe the HTTP verb work should be dealt with in the config/routes.rb and not here?

darrenleeweber replied:
Yes, good point, and it is dealt with in the routes, so the system should never get inside this `unless` block.  However, for completeness, the IIIF spec has requirements for the response content and this code block takes care of that.  If indeed, the routing ensures this code block is never hit, then @azaroth42 might want to comment on whether we rely on the default rails handlers/responses or we must have this controller handle the response to ensure the response body conforms to spec?

azaroth42 replied:
Sticking to what other developers will expect is the best, IMO.  My understanding is that routes is where those things would normally be, so if (and I have no idea why) anyone would need to find where the handling was, that's probably where they'd look first.  If there's no possibility for the code to be called ... then I'd delete it as unnecessary and just adding extra maintenance.

ndushay commented on bin/ctags.rb:
```diff
@@ -0,0 +1,8 @@
+#!/usr/bin/env ruby
```
I'm not certain your bin/ctags.rb  should go in here - I think you're pretty unique in DLSS and Hydra community in caring about ctags ...

darrenleeweber replied:
Thanks for pointing that out, I didn't really notice it.  It does no harm, but I can either remove it and modify the .gitignore, or I can leave it and modify code so it doesn't get into a gem.

ndushay commented on config/routes.rb:
```diff
@@ -8,75 +16,58 @@
-      anno_root = request.path_parameters[:anno_root]
-      anno_root !~ /^annotations$/ && anno_root !~ /^search$/ && anno_root !~ /^new$/
-    }
+    constraints: lambda { |request| anno_container_exclusions request }
```
your solution is adding id exclusions here

ndushay commented on config/routes.rb:
```diff
@@ -8,75 +16,58 @@
 
 end
+
+
```
perhaps it would be more clear above if id_exclusions was not part of container_exclusions.   I think it's fine to exclude "annotations" as an auth_root value for everything.

darrenleeweber replied:
These points should be addressed in the commit today (788486e).

ndushay commented on spec/controllers/triannon/auth_controller_spec.rb:
```diff
@@ -0,0 +1,390 @@
+    request.headers['Content-Type'] =  'application/json'
+  }
+
+  describe '/auth/login' do
```
the convention is for routing specs to live in  spec/routing

darrenleeweber replied:
Thanks, all the routing specs have been adjusted in 678978a.

ndushay commented on spec/controllers/triannon/auth_controller_spec.rb:
```diff
@@ -0,0 +1,390 @@
+  end # /auth/login
+
+  describe 'GET /auth/logout' do
+    describe 'only responds to GET requests' do
```
more routing specs ...

ndushay commented on spec/controllers/triannon/auth_controller_spec.rb:
```diff
@@ -0,0 +1,390 @@
+    before :each do
+      basic_auth('userA', 'secretA')
+    end
+    describe 'only responds to POST requests' do
```
... routing specs ...

ndushay commented on spec/controllers/triannon/auth_controller_spec.rb:
```diff
@@ -0,0 +1,390 @@
+      expect(data['accessToken']).not_to be_nil
+      expect(data['accessToken']).to be_instance_of String
+    }
+    describe 'only responds to GET requests' do
```
... routing specs ...

ndushay commented on .gitattributes:
```diff
@@ -0,0 +1,2 @@
+spec/fixtures/vcr_cassettes -diff
```
sadly, it doesn't look like this affected the diff showing here in github "files changed"

darrenleeweber replied:
Maybe that is so, or maybe it's because the PR was started before adding this.  It may be that the PR is based on `merge` rather than `diff` data, maybe the next PR will be more telling.

ndushay commented:
Overall:  I can't offer anything useful on the guts of the actual authentication code;  the auth code functionality seems okay to my ignorant eyeball overview.

My main concerns:
  auth timeouts should be a config setting, not hardcoded
  HTTP verb handling should be done with routes, not in controller, if possible
  routing specs are conventionally put in spec/routing, not made part of controller specs.

I don't think the .gitattributes thing worked as the vcr diffs are still showing in this PR.  Too bad.

BTW, rails routes are really sensitive to ordering, so it *might* have worked to put your auth routes at the end, but I'm fine with your routing changes (caveat:  need the fix to not do id constraints when not needed)

darrenleeweber commented:
I think the coveralls failure is trivial, so this PR should now have addressed most of the points raised in @ndushay's review.  Additional points require consideration and advice from @azaroth42.

darrenleeweber commented:
@azaroth42 - please review the auth-code, as @ndushay is unfamiliar with the IIIF auth-spec.  For reference, see https://github.com/IIIF/auth/blob/master/server/pi3f.oauth.py

azaroth42 commented:
It looks correct to me, but NB I'm just following the high level logic, not the code itself.

darrenleeweber commented:
In the interests of moving on, I'll merge this and address additional issues in the `auth-ish` branch, once these changed are merged/integrated there.
